### PR TITLE
Docs: build with Sphinx 5

### LINKED ISSUE #60979: py:class reference target not found: torch.utils.data.dataset.Dataset
## 📚 Documentation

I have a project that uses PyTorch. When creating my own documentation, I'm using [intersphinx](https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html) to automatically link to the PyTorch documentation like so:
```python
intersphinx_mapping = {                                                                  
    "python": ("https://docs.python.org/3", None),                                       
    "torch": ("https://pytorch.org/docs/stable", None),                                  
}  
```
When I build my project documentation with nitpick mode enabled, sphinx warns that it can't find certain class definitions:
```
foo.py:docstring of foo.Foo:1: WARNING: py:class reference target not found: torch.utils.data.dataset.Dataset
foo.py:docstring of foo.Bar:1: WARNING: py:class reference target not found: torch.nn.modules.module.Module
```
This usually implies that there is something wrong with the `objects.inv` file available at https://pytorch.org/docs/stable/objects.inv

I could create my own `objects.inv` for local use, but it would quickly become out-of-date. Any idea how we can fix this in master?

### PR DESCRIPTION
**Pull request by adamjstewart**
Fixes #60979. Also see #61045 and https://github.com/sphinx-doc/sphinx/issues/9395 for discussion.

I _believe_ the reason that we were previously pinning to Sphinx 3 was because of issues with pytorch_sphinx_theme and Sphinx 4 support, but these seem to have been resolved now. See https://torchgeo.readthedocs.io/ for an example of docs built with pytorch_sphinx_theme and Sphinx 4.

### CODE REVIEW

**Comment by samdow** [comment]
@brianjo Adding you as reviewer from issue discussion but feel free to send it to someone else who knows this area better

**Comment by adamjstewart** [comment]
@branjo can you review this? Also pinging @mruberry @ydcjeff @mattip who commented on the issue/PR that this references.

File: docs/source/torch.rst [hunk 1]
```diff
@@ -263,11 +263,8 @@ Examples::
     :toctree: generated
     :nosignatures:
 
-    no_grad
```
Comment (mruberry) [inline docs/source/torch.rst:266]: Hey @adamjstewart! I'll let @brianjo look at the Sphinx update, but why does this PR remove these entries?

Comment (adamjstewart) [reply docs/source/torch.rst:266]: Sphinx raises a warning when entries in the index appear in multiple places in the docs. I believe this was always intended to be the case, but Sphinx 4 fixed a bug that prevented it from being detected properly. See the errors raised by previous commits. The error messages suggest adding `:noindex:` to one of the entries but that didn't seem to work for me. I'm guessing this has something to do with things being auto-generated.

TL;DR this is a hack to get the documentation tests to pass. I welcome any suggestions of alternative ways to get the tests to pass but ran out of ideas myself.

Comment (mruberry) [reply docs/source/torch.rst:266]: Interesting; thanks for the explanation. Let's wait for our docs expert(s) to weigh-in.

Comment (mattip) [reply docs/source/torch.rst:266]: Is there a way to see the rendered documentation after this PR? I would expect the removed entries to still show up like they do on https://pytorch.org/docs/stable/torch.html#locally-disabling-gradient-computation as rows in the table below the examples.

**Comment by adamjstewart** [comment]
@brianjo have you gotten a chance to look at this PR yet?

File: docs/source/torch.rst (hunk 1 above)
Comment (brianjo) [reply docs/source/torch.rst:266]: @seemethere Eli, is there a way for us to get the output of the docs build now that we're on actions?

Comment (NicolasHug) [reply docs/source/torch.rst:266]: The rendered docs are available from the GithubBot post: https://github.com/pytorch/pytorch/pull/70309#issuecomment-999723231


https://docs-preview.pytorch.org/70309/index.html

**Comment by adamjstewart** [comment]
@brianjo do these changes look good to you?

File: docs/requirements.txt [hunk 2]
```diff
@@ -1,4 +1,4 @@
-sphinx==3.5.4
+sphinx==4.3.2
```
Comment (adamjstewart) [inline docs/requirements.txt:1]: Sphinx is already on their 4.4 release. Maybe we shouldn't pin this at all?

**Comment by adamjstewart** [comment]
Ping @brianjo

**Comment by adamjstewart** [comment]
Is anyone willing to review this PR? This PR has been stagnant for a few months now...

**Comment by malfet** [comment]
@adamjstewart can you please rebase the code, and if doc build job is fine then it's good to go.

**Review by malfet — APPROVED** [review]

File: docs/requirements.txt (hunk 2 above)
Comment (malfet) [reply docs/requirements.txt:1]: Pinning is good, otherwise updates might break build pipeline

**Review by malfet — CHANGES_REQUESTED** [review]
Requesting change as after the rebase `no_grad` option is missing from newly generated docs

Compare https://docs-preview.pytorch.org/70309/torch.html#locally-disabling-gradient-computation with stable doc:
image

File: docs/source/torch.rst (hunk 1 above)
Comment (malfet) [reply docs/source/torch.rst:269]: @adamjstewart your change removes number of entries from the doc, compare:
https://pytorch.org/docs/stable/torch.html#locally-disabling-gradient-computation
with
https://docs-preview.pytorch.org/70309/torch.html#locally-disabling-gradient-computation

`no_grad` is now missing in the preview

Comment (adamjstewart) [reply docs/source/torch.rst:269]: Correct, see my comment above: https://github.com/pytorch/pytorch/pull/70309#discussion_r782322575

If anyone knows how to fix this without removing `no_grad` from the docs please let me know, but I'm out of ideas. The current docs are not valid as far as Sphinx is concerned, you aren't allowed to have the same entry in the index multiple times.

Comment (malfet) [reply docs/source/torch.rst:262]: I'm not sure, I understand the comment: why do you think `no_grad` index appears in multiple places? Greping thru [torch.rst](https://github.com/pytorch/pytorch/blob/master/docs/source/torch.rst) shows it mentioned only once in that .rst file, though I'm not at all a doc expert

Comment (adamjstewart) [reply docs/source/torch.rst:262]: It's also mentioned in autograd.rst. All files are combined to create a single index for the documentation.

**Comment by adamjstewart** [comment]
I reverted the changes to `torch.rst` to help illustrate the issue and error message that arises from having the same function appear multiple times in the index.

**Comment by adamjstewart** [comment]
Okay, I reached out to a friend who knows a lot more about Sphinx than I ever will. He did a really good job of explaining what's going on here, so I'll try to summarize it below.

The root of the issue is that several methods are actually documented twice in the documentation.

`docs/source/torch.rst`:
```rst
.. autosummary::
    :toctree: generated
    :nosignatures:

    no_grad
    enable_grad
    set_grad_enabled
    is_grad_enabled
    inference_mode
    is_inference_mode_enabled
```
`docs/source/autograd.rst`:
```rst
.. autosummary::
    :toctree: generated
    :nosignatures:

    no_grad
    enable_grad
    set_grad_enabled
    inference_mode
```
The [autosummary](https://www.sphinx-doc.org/en/master/usage/extensions/autosummary.html) directive generates a pretty table like you see [here](https://pytorch.org/docs/stable/torch.html#locally-disabling-gradient-computation) and [here](https://pytorch.org/docs/stable/autograd.html#locally-disabling-gradient-computation). Since we use `:toctree: generated` and have `autosummary_generate = True` in `conf.py`, it also generates separate pages like you see [here](https://pytorch.org/docs/stable/generated/torch.no_grad.html#torch.no_grad) and [here](https://pytorch.org/docs/stable/generated/torch.autograd.no_grad.html#torch.autograd.no_grad).

The problem is that Sphinx explicitly disallows you from documenting the same method in multiple locations in your documentation. This is important, for example, for interpshinx since it becomes ambiguous which method to link to. The fact that the documentation builds at all with Sphinx 3 [is a bug](https://github.com/sphinx-doc/sphinx/issues/7817) that was fixed in Sphinx 4, hence the errors we see when building the docs with modern Sphinx.

The error message itself recommends using `:noindex:` on one of the locations to avoid adding it to the index. Autosummary is simply syntactic sugar for a bunch of [autodoc](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html) calls. All autodoc methods support `:noindex:`, but autosummary itself does not. That's why my previous commits didn't fix the issue.

So the _correct_ solution is to not document these methods twice. The last couple of commits are an attempt to hack Sphinx to allow this anyway, but I haven't gotten it working yet, and it really isn't recommended. So I think someone needs to make an executive decision here. It's not really possible to keep the docs the way they are now and use modern versions of Sphinx that fix issues like #60979. I strongly recommend removing the duplicate method documentation in one location or the other.

**Comment by NicolasHug** [comment]
> All autodoc methods support :noindex:, but autosummary itself does not

You can tell autosummary to use a template (this one is used a lot in the code-base apparently: https://github.com/pytorch/pytorch/blob/master/docs/source/_templates/classtemplate.rst). You should be able to put the `:noindex:` in a new custom template?

Or better,  make the template more generic and accept a "noindex" parameter (but I don't know if that's possible).

**Comment by adamjstewart** [comment]
I think the problem with the template approach is that this would affect all members, not just the ones that are documented twice. Let me see if I can get the `conf.py` hack working...

**Comment by adamjstewart** [comment]
I think I found a good compromise! The latest commit removes ":toctree: generated" from only `autograd.rst`, where all members are already documented in `torch.rst` and are therefore duplicates. Both locations will still have the pretty autosummary table, but pages like [torch.autograd.no_grad.html](https://pytorch.org/docs/stable/generated/torch.autograd.no_grad.html#torch.autograd.no_grad) will no longer exist, there will only be [torch.no_grad.html](https://pytorch.org/docs/stable/generated/torch.no_grad.html#torch.no_grad).

**Comment by adamjstewart** [comment]
If you want I can just remove the table...

Seriously though, I've tried just about everything I can think of to make this work, but Sphinx doesn't let you document things twice and I'm running out of ideas. I would love to hear suggestions for how to solve this bug.

**Comment by albanD** [comment]
Are these 4 functions the only problematic ones? Do we have a list for these?

**Comment by NicolasHug** [comment]
TBH at this point I would just remove these functions from the `.. autosummary` in the `torch.rst` page. All of them are already mentioned and clickable [just above](https://github.com/pytorch/pytorch/blob/master/docs/source/torch.rst#locally-disabling-gradient-computation), except for `torch.is_grad_enabled`, but we can add it as well.

The remaining autosummary table in torch.rst would just contain `inference_mode` and `is_inference_mode_enabled`.

**Comment by albanD** [comment]
Sounds good to me!

**Comment by adamjstewart** [comment]
Alright, I removed the duplicate entries from `torch.rst` as suggested. It looks like Sphinx 5 was just released so I'm trying that too. Hopefully that doesn't add any new surprises.

File: docs/requirements.txt [hunk 3]
```diff
@@ -1,6 +1,4 @@
-sphinx==3.5.4
-Jinja2==3.0.*
-docutils==0.16
```
Comment (adamjstewart) [inline docs/requirements.txt:3]: Jinja2 and docutils are both Sphinx dependencies, we aren't using them directly as far as I know. They were only pinned because Sphinx 3 doesn't work with newer Jinja2 and Sphinx 3–4 don't work with newer docutils. But the latest versions of both should work with Sphinx 5, so there's no longer a reason to pin them.

File: docs/source/torch.rst [hunk 4]
```diff
@@ -260,11 +260,7 @@ Examples::
...
-    enable_grad
-    set_grad_enabled
```
Comment (albanD) [inline docs/source/torch.rst:265]: I think the text just above does refer to these functions. Maybe removing the ones in `autograd.rst` would be less disruptive? We can also add plain text there that links to the `torch.` version of these docs.

File: docs/requirements.txt (hunk 3 above)
Comment (adamjstewart) [reply docs/requirements.txt:3]: I'm also happy to unpin Sphinx and always test with the latest version, but I know you mentioned wanting more CI stability. None of the other deps are pinned though. Personally, I would only pin things if this file was managed with dependabot and received regular updates.

Comment (albanD) [reply docs/requirements.txt:3]: Just pinning sphinx should be ok yes.

File: docs/source/torch.rst (hunk 4 above)
Comment (adamjstewart) [reply docs/source/torch.rst:265]: That's what I originally did. But I like @NicolasHug's comment:

> All of them are already mentioned and clickable [just above](https://github.com/pytorch/pytorch/blob/master/docs/source/torch.rst#locally-disabling-gradient-computation), except for `torch.is_grad_enabled`, but we can add it as well.

So the idea is that the table remains in `autograd.rst` but `torch.rst` still has links to the same docs in approximately the same location.

I'm happy to remove whichever table entries you decide on though.

File: docs/requirements.txt (hunk 3 above)
Comment (malfet) [reply docs/requirements.txt:3]: Let's keep version pinned, otherwise CI/CD might run into a problem every time new version of dependent package is released, which happened  just few days ago with release of protobuf-4.23, see https://github.com/pytorch/pytorch/issues/78362

File: docs/source/torch.rst (hunk 4 above)
Comment (albanD) [reply docs/source/torch.rst:265]: > All of them are already mentioned and clickable [just above](https://github.com/pytorch/pytorch/blob/master/docs/source/torch.rst#locally-disabling-gradient-computation), except for torch.is_grad_enabled, but we can add it as well.

But they are clickable only because we still have entries for `torch.` versions. If we only keep the `torch.autograd.` versions, these won't be clickable anymore right?

File: docs/source/torch.rst (hunk 1 above)
Comment (malfet) [reply docs/source/torch.rst:263]: @adamjstewart I can open it in incognito browser, so I always thought that they are public

File: docs/requirements.txt (hunk 3 above)
Comment (adamjstewart) [reply docs/requirements.txt:3]: Heh, protobuf was fun.

File: docs/source/torch.rst (hunk 4 above)
Comment (albanD) [reply docs/source/torch.rst:265]: The `torch.no_grad()` usage is definitely more common. That's why I was suggesting removing the `autograd.rst` table. We can just replace that table with a link to the `torch.rst` paragraph that explain these functions.

Comment (adamjstewart) [reply docs/source/torch.rst]: Done

**Review by albanD — COMMENTED** [review]
SGTM
Let's check the rendered version to make sure the 5.0 upgrade didn't break anything rendering-wise.

**Review by albanD — APPROVED** [review]
The code sample looks quite a bit better as well.
Thanks!

**Comment by adamjstewart** [comment]
Thanks for the help on this everyone!
